src/ports.rs

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use enarx_config::{Config, File, Protocol};
+use lazy_static::lazy_static;
+use tokio::sync::{RwLock, RwLockWriteGuard};
+
+lazy_static! {
+    static ref PORTS_IN_USE: Arc<RwLock<Vec<u16>>> = Arc::new(RwLock::new(vec![]));
+}
+
+pub fn get_listen_ports(toml: &str) -> anyhow::Result<Vec<u16>> {
+    let config = toml::from_str::<Config>(toml).with_context(|| "failed to parse enarx config")?;
+
+    let ports = config
+        .files
+        .into_iter()
+        .filter_map(|file| match file {
+            File::Null { .. }
+            | File::Stdin { .. }
+            | File::Stdout { .. }
+            | File::Stderr { .. }
+            | File::Connect { .. } => None,
+            File::Listen { port, prot, .. } => match prot {
+                Protocol::Tls | Protocol::Tcp => Some(port),
+            },
+        })
+        .collect();
+
+    Ok(ports)
+}
+
+pub async fn try_reserve(ports: &[u16]) -> Result<(), Vec<u16>> {
+    let mut ports_in_use = PORTS_IN_USE.write().await;
+    let conflicting = conflicting(ports, Some(&ports_in_use)).await;
+
+    if !conflicting.is_empty() {
+        return Err(conflicting);
+    }
+
+    ports_in_use.extend(ports);
+    Ok(())
+}
+
+pub async fn free(ports: &[u16]) {
+    let mut ports_in_use = PORTS_IN_USE.write().await;
+    let to_remove = ports_in_use
+        .iter()
+        .enumerate()
+        .filter_map(|(i, port)| if ports.contains(port) { Some(i) } else { None })
+        .collect::<Vec<_>>();
+
+    for index in to_remove {
+        ports_in_use.remove(index);
+    }
```

Comment:
if we used a `HashSet`, we could simply loop over the ports and remove each from the set.
In fact, this function would probably be redundant, since it's a trivial operation